File: nebras-app/app/src/main/kotlin/com/nebras/mobile/feature/auth/SignInScreen.kt
package com.nebras.mobile.feature.auth

import androidx.activity.compose.BackHandler
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.nebras.mobile.core.theme.NebrasColors
import com.nebras.mobile.core.theme.NebrasTextStyles
import com.nebras.mobile.core.util.ar
import kotlinx.coroutines.launch

/**
 * شاشة الدخول/التسجيل بحساب Google.
 *
 * نستخدم **نفس الشاشة** لحالتَي «إنشاء حساب» و«تسجيل الدخول»، ونُمرّر
 * [isSignUp] فقط لتغيير النصوص: مُنتقي حسابات Google هو نفسه تقنيّاً
 * (لا فرق في الـ API)، وهذا يمنح تجربة موحّدة ويقلّل التكرار.
 */
@Composable
fun SignInScreen(
    isSignUp: Boolean,
    authViewModel: AuthViewModel,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val auth by authViewModel.state.collectAsStateWithLifecycle()
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val snackbarHostState = remember { SnackbarHostState() }

    // زرّ الرجوع الماديّ يعادل pop في Flutter.
    BackHandler(onBack = onBack)

    Scaffold(
        modifier = modifier,
        containerColor = MaterialTheme.colorScheme.background,
        snackbarHost = { SnackbarHost(snackbarHostState) },
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .padding(horizontal = 28.dp),
        ) {
            // شريط علويّ شفّاف بزرّ رجوع فقط (مقابل AppBar الفارغ في Flutter).
            Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                IconButton(onClick = onBack) {
                    Icon(
                        Icons.AutoMirrored.Filled.ArrowBack,
                        contentDescription = "رجوع",
                        tint = MaterialTheme.colorScheme.onSurface,
                    )
                }
            }

            Spacer(Modifier.height(16.dp))
            Text(
                text = if (isSignUp) "إنشاء حسابك" else "مرحباً بعودتك",
                style = NebrasTextStyles.heading1,
            )
            Spacer(Modifier.height(10.dp))
            Text(
                text = if (isSignUp) {
                    "اختر حساب Google من هذا الجهاز لإنشاء حساب نبراس الخاص بك " +
                        "والاستمتاع بمحتوى مخصّص."
                } else {
                    "اختر حساب Google المستخدم سابقاً للدخول إلى نبراس واستعادة بياناتك."
                },
                // height: 1.6 → 16sp × 1.6
                style = NebrasTextStyles.body.copy(
                    color = NebrasColors.textSecondary,
                    lineHeight = 25.6.sp,
                ),
            )

            Spacer(Modifier.weight(1f))

            GoogleButton(
                busy = auth.isBusy,
                label = if (isSignUp) "التسجيل عبر Google" else "الدخول عبر Google",
                enabled = !auth.isBusy,
                onPressed = {
                    scope.launch {
                        val success = authViewModel.signInWithGoogle(context)
                        if (success) {
                            // نجاح الدخول: `AuthGate` غيّر حالته إلى signedIn
                            // ويعرض الواجهة الرئيسية، فنُغلق شاشة الدخول كي
                            // لا تبقى معلّقة فوقها.
                            onBack()
                            return@launch
                        }
                        val err = authViewModel.state.value.lastError
                        if (err != null) {
                            // المفاتيح التي يُرجعها AuthViewModel تُحلّ نصّاً
                            // عربيّاً عبر `.ar`؛ وإن لم يُطابق المفتاح أيّ
                            // إدخال يُعاد النصّ نفسه كما هو.
                            snackbarHostState.showSnackbar(err.ar)
                            authViewModel.clearError()
                        }
                    }
                },
            )
            Spacer(Modifier.height(20.dp))
            Text(
                text = "نستخدم حساب Google للتعرّف على هويتك فقط. لن نتلقّى كلمة المرور " +
                    "ولن نشارك أي بيانات شخصية.",
                modifier = Modifier.fillMaxWidth(),
                textAlign = TextAlign.Center,
                // height: 1.5 → 14sp × 1.5
                style = NebrasTextStyles.greysmallText.copy(lineHeight = 21.sp),
            )
            Spacer(Modifier.height(24.dp))
        }
    }
}

@Composable
private fun GoogleButton(
    busy: Boolean,
    label: String,
    enabled: Boolean,
    onPressed: () -> Unit,
) {
    Button(
        onClick = onPressed,
        enabled = enabled,
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp),
        shape = RoundedCornerShape(14.dp),
        border = androidx.compose.foundation.BorderStroke(1.dp, NebrasColors.border),
        colors = ButtonDefaults.buttonColors(
            containerColor = Color.White,
            contentColor = NebrasColors.textPrimary,
            disabledContainerColor = Color.White,
            disabledContentColor = NebrasColors.textPrimary,
        ),
        elevation = ButtonDefaults.buttonElevation(defaultElevation = 1.5.dp),
    ) {
        if (busy) {
            CircularProgressIndicator(
                modifier = Modifier.size(22.dp),
                strokeWidth = 2.2.dp,
                color = NebrasColors.primary,
            )
        } else {
            Row(
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                GoogleLogo()
                Spacer(Modifier.width(12.dp))
                Text(
                    text = label,
                    style = NebrasTextStyles.bodyBold.copy(color = NebrasColors.textPrimary),
                )
            }
        }
    }
}

/**
 * شعار Google مرسوم بخطّ بسيط بدلاً من أيقونة شبكيّة أو أصل إضافي.
 * نعتمد أربع ألوان العلامة الرسمية (نقل `_GooglePainter`).
 */
@Composable
private fun GoogleLogo() {
    val blue = Color(0xFF4285F4)
    val red = Color(0xFFEA4335)
    val yellow = Color(0xFFFBBC05)
    val green = Color(0xFF34A853)

    Box(Modifier.size(22.dp)) {
        Canvas(Modifier.fillMaxSize()) {
            val w = size.width
            val center = Offset(w / 2f, size.height / 2f)
            val radius = w / 2f
            val arcRadius = radius * 0.95f
            val topLeft = Offset(center.x - arcRadius, center.y - arcRadius)
            val arcSize = Size(arcRadius * 2f, arcRadius * 2f)
            val stroke = Stroke(width = w * 0.22f, cap = StrokeCap.Butt)

            // 4 أقواس لتقليد الشعار الرسمي (مبسّطة للبنية الرياضية).
            // زوايا Flutter بالراديان → درجات (Compose يستعمل الدرجات).
            val rad = (180.0 / Math.PI).toFloat()
            drawArc(blue, -1.05f * rad, 1.15f * rad, false, topLeft, arcSize, style = stroke)
            drawArc(green, 0.10f * rad, 1.30f * rad, false, topLeft, arcSize, style = stroke)
            drawArc(yellow, 1.40f * rad, 1.35f * rad, false, topLeft, arcSize, style = stroke)
            drawArc(red, 2.75f * rad, 1.25f * rad, false, topLeft, arcSize, style = stroke)

            // شريط G الداخلي (خطّ أزرق أفقي صغير).
            val barWidth = w * 0.38f
            val barHeight = w * 0.18f
            val barCenter = Offset(center.x + w * 0.18f, center.y)
            drawRect(
                color = blue,
                topLeft = Offset(barCenter.x - barWidth / 2f, barCenter.y - barHeight / 2f),
                size = Size(barWidth, barHeight),
            )
        }
    }
}

File: app-flutter/lib/features/auth/view/sign_in_screen.dart
import 'package:easy_localization/easy_localization.dart';
import 'package:flutter/material.dart';
import 'package:flutter_screenutil/flutter_screenutil.dart';
import 'package:nebras_mobile_app/core/extensions/sized_boxextension.dart';
import 'package:nebras_mobile_app/core/theme/color_manager.dart';
import 'package:nebras_mobile_app/core/theme/text_styles.dart';
import 'package:nebras_mobile_app/features/auth/provider/auth_provider.dart';
import 'package:provider/provider.dart';

/// شاشة الدخول/التسجيل بحساب Google.
///
/// نستخدم **نفس الشاشة** لحالتَي "إنشاء حساب" و"تسجيل الدخول"، ونُمرّر
/// `isSignUp` فقط لتغيير النصوص. السبب:
///   • مُنتقي حسابات Google هو نفسه تقنيًا (لا فرق في الـ API).
///   • يمنح ذلك تجربة موحّدة ومتسقّة، ويقلّل التكرار.
///   • مستقبلاً، يمكن للمطوّر التفريق بين حالة "مستخدم جديد" و"عائد" عبر
///     علامة داخليّة (first-seen) بعد نجاح الدخول.
class SignInScreen extends StatelessWidget {
  final bool isSignUp;
  const SignInScreen({super.key, required this.isSignUp});

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final auth = context.watch<AuthProvider>();

    return Scaffold(
      backgroundColor: theme.scaffoldBackgroundColor,
      appBar: AppBar(
        backgroundColor: Colors.transparent,
        elevation: 0,
        iconTheme: IconThemeData(color: theme.iconTheme.color),
      ),
      body: SafeArea(
        child: Padding(
          padding: EdgeInsets.symmetric(horizontal: 28.w),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              16.sbh,
              Text(
                isSignUp ? 'Create your account'.tr() : 'Welcome back'.tr(),
                style: TextStyles.heading1(context),
              ),
              10.sbh,
              Text(
                isSignUp
                    ? 'Signup subtitle'.tr()
                    : 'Signin subtitle'.tr(),
                style: TextStyles.body(context).copyWith(
                  color: ColorsManager.textSecondary,
                  height: 1.6,
                ),
              ),
              const Spacer(),
              _GoogleButton(
                busy: auth.isBusy,
                label: isSignUp
                    ? 'Sign up with Google'.tr()
                    : 'Sign in with Google'.tr(),
                onPressed: auth.isBusy
                    ? null
                    : () => _handleGoogle(context),
              ),
              20.sbh,
              Text(
                'Privacy note'.tr(),
                textAlign: TextAlign.center,
                style: TextStyles.greysmallText(context).copyWith(height: 1.5),
              ),
              24.sbh,
            ],
          ),
        ),
      ),
    );
  }

  Future<void> _handleGoogle(BuildContext context) async {
    final auth = context.read<AuthProvider>();
    final success = await auth.signInWithGoogle();
    if (!context.mounted) return;
    if (success) {
      // بعد نجاح الدخول: `AuthGate` غيّر حالته إلى `signedIn` وبات يعرض
      // `HomeShell` تحتنا مباشرةً، لكنّ `SignInScreen` (وأحيانًا `WelcomeScreen`
      // إن وُجدت قبلها) لا تزال مدفوعة في الـ root Navigator فوق `AuthGate`.
      // لذا نُرجِع الملاحة إلى الجذر — أيّ نُغلق كلّ شيء فوق `AuthGate` —
      // فيرى المستخدم الصفحة الرئيسية فورًا دون الحاجة للضغط على "رجوع".
      // `popUntil(isFirst)` آمن حتى لو لم تكن هناك شاشات فوقنا (لا يفعل شيئًا).
      Navigator.of(context).popUntil((route) => route.isFirst);
      return;
    }
    final err = auth.lastError;
    if (err != null) {
      // المفاتيح التي يُرجعها AuthProvider قابلة للترجمة مباشرة عبر .tr().
      // إن لم يُطابق المفتاح أيّ إدخال (حالة رسالة فنّية جاءت من الحزمة)،
      // يُعيد easy_localization نفس النص كما هو.
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(
          content: Text(
            err.tr(),
            style: TextStyles.smallText(context).copyWith(color: Colors.white),
          ),
          backgroundColor: ColorsManager.error,
          behavior: SnackBarBehavior.floating,
          margin: EdgeInsets.all(16.w),
        ),
      );
      auth.clearError();
    }
  }
}

class _GoogleButton extends StatelessWidget {
  final bool busy;
  final String label;
  final VoidCallback? onPressed;

  const _GoogleButton({
    required this.busy,
    required this.label,
    required this.onPressed,
  });

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: 56.h,
      child: ElevatedButton(
        onPressed: onPressed,
        style: ElevatedButton.styleFrom(
          backgroundColor: Colors.white,
          foregroundColor: ColorsManager.textPrimary,
          elevation: 1.5,
          shadowColor: Colors.black.withValues(alpha: 0.08),
          side: const BorderSide(color: ColorsManager.border),
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(14.r),
          ),
        ),
        child: busy
            ? SizedBox(
                width: 22.w,
                height: 22.w,
                child: const CircularProgressIndicator(
                  strokeWidth: 2.2,
                  color: ColorsManager.primary,
                ),
              )
            : Row(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  const _GoogleLogo(),
                  12.sbw,
                  Text(
                    label,
                    style: TextStyles.bodyBold(context).copyWith(
                      color: ColorsManager.textPrimary,
                    ),
                  ),
                ],
              ),
      ),
    );
  }
}

/// شعار Google مرسوم بخطّ بسيط بدلاً من أيقونة شبكيّة أو أصل إضافي.
/// نعتمد أربع ألوان العلامة الرسمية.
class _GoogleLogo extends StatelessWidget {
  const _GoogleLogo();

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: 22.w,
      height: 22.w,
      child: CustomPaint(painter: _GooglePainter()),
    );
  }
}

class _GooglePainter extends CustomPainter {
  static const _blue = Color(0xFF4285F4);
  static const _red = Color(0xFFEA4335);
  static const _yellow = Color(0xFFFBBC05);
  static const _green = Color(0xFF34A853);

  @override
  void paint(Canvas canvas, Size size) {
    final center = size.center(Offset.zero);
    final radius = size.width / 2;
    final rect = Rect.fromCircle(center: center, radius: radius * 0.95);
    final paint = Paint()
      ..style = PaintingStyle.stroke
      ..strokeWidth = size.width * 0.22
      ..strokeCap = StrokeCap.butt;

    // 4 أقواس لتقليد الشعار الرسمي (مبسّطة للبنية الرياضية).
    canvas.drawArc(rect, -1.05, 1.15, false, paint..color = _blue);
    canvas.drawArc(rect, 0.10, 1.30, false, paint..color = _green);
    canvas.drawArc(rect, 1.40, 1.35, false, paint..color = _yellow);
    canvas.drawArc(rect, 2.75, 1.25, false, paint..color = _red);

    // شريط G الداخلي (خطّ أزرق أفقي صغير).
    final bar = Rect.fromCenter(
      center: center.translate(size.width * 0.18, 0),
      width: size.width * 0.38,
      height: size.width * 0.18,
    );
    canvas.drawRect(
      bar,
      Paint()..color = _blue,
    );
  }

  @override
  bool shouldRepaint(covariant _GooglePainter oldDelegate) => false;
}
